### ISSUE TITLE
Hardhat node OOM when build-info file is big

### ISSUE THREAD

**Issue by jlcs-es** [NomicFoundation/hardhat#3346]
## Description

When the file in `build-info/` is too big (in our case ~263M) hardhat node crashes after a few seconds without output. Even with a heap size of 12GB [1]. This actually happened during `hardhat test`, **without any logs indicating this is a hardhat node's issue**.

Deleting the file inside `build-info/` allows `npx hardhat node` to run.

Another project has a `build-info` file around 100MB and hardhat node works, but it has a **notable delay** (~5s) compared with launching after deleting the file.

<details>
  <summary>[1] OOM console output</summary>
  

```
$ NODE_OPTIONS="--max-old-space-size=12000" npx hardhat node

<--- Last few GCs --->

[432788:0x54c4e60]    62007 ms: Mark-sweep 11816.8 (12032.6) -> 11811.5 (12032.6) MB, 3509.3 / 0.0 ms  (average mu = 0.331, current mu = 0.083) allocation failure scavenge might not succeed
[432788:0x54c4e60]    68234 ms: Mark-sweep 11828.6 (12032.9) -> 11828.5 (12060.4) MB, 6221.0 / 0.0 ms  (average mu = 0.181, current mu = 0.001) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb0a860 node::Abort() [node]
 2: 0xa1c193 node::FatalError(char const*, char const*) [node]
 3: 0xcf9a6e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xcf9de7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xeb1685  [node]
 6: 0xec134d v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0xec404e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0xe8558a v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
 9: 0x11fe2d6 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
10: 0x15f2d39  [node]
Aborted (core dumped)
```
</details>

## Related issues

- #1186
- #2803

## Quick-Fix questions

What are the drawbacks of deleting `build-info/` between compilation and tests?
Why would `hardhat node` need to load that info?

**Comment by fvictorio** [comment]
> What are the drawbacks of deleting `build-info/` between compilation and tests?
> Why would `hardhat node` need to load that info?

Hardhat uses the compilation output to generate stack traces and infer the cause of reverted transactions. You can delete them and run `npx hardhat test --no-compile` and your tests should work fine, but if something fails you'll get less information.

Thanks for reporting this, it seems important.

**Comment by fvictorio** [comment]
Hey @jlcs-es, is your project public? I would like to investigate this a bit.

**Comment by jlcs-es** [comment]
@fvictorio sadly no, but you can generate a contract doing something like this (tested locally with 8GB of RAM and hardhat v2.12.4):
- Create new contract `A.sol` with:

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.7.0 <0.9.0;
contract A {
    string bigasstext = "
```
> No new line at the end!


- In terminal generate a lot of random text in a single line (change `count=` to adapt size):
```bash
random=$(dd if=/dev/urandom bs=1 count=100000000 | tr -dc 'a-zA-Z0-9~!@#$%^&*_-' | base64 -w 0)
```
> This may take a while...


- Then append at the end without new lines:
```bash
printf "%s%s%s" "$(cat contracts/A.sol)" "$random" "\"; }" > contracts/A.sol
```

- Compile, check size and try to run hh node:

```bash
$ npx hardhat compile
Compiled 1 Solidity file successfully

$ du -h artifacts/build-info/
413M	artifacts/build-info/

$ npx hardhat node

<--- Last few GCs --->

[32438:0x6ed9130]    33131 ms: Scavenge 2022.8 (2053.5) -> 2022.8 (2058.3) MB, 8.0 / 0.0 ms  (average mu = 0.176, current mu = 0.153) allocation failure; 
[32438:0x6ed9130]    33146 ms: Scavenge 2025.9 (2058.3) -> 2026.2 (2059.8) MB, 9.8 / 0.0 ms  (average mu = 0.176, current mu = 0.153) allocation failure; 
[32438:0x6ed9130]    33159 ms: Scavenge 2026.9 (2059.8) -> 2026.6 (2070.8) MB, 12.1 / 0.0 ms  (average mu = 0.176, current mu = 0.153) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb6e500 node::Abort() [node]
[...]
```

**Comment by fvictorio** [comment]
@jlcs-es thanks for the detailed steps, but I still can't reproduce it. Are you using the latest version of Hardhat? We made a change to how the build info files are generated, which might have fixed this issue.

**Comment by jlcs-es** [comment]
@fvictorio I have checked my node and hh versions and recorded muy terminal session for faster demo:

[![asciicast](https://asciinema.org/a/LVNiA4sdp0lb5iJFpcer8HB1j.svg)](https://asciinema.org/a/LVNiA4sdp0lb5iJFpcer8HB1j)

**Comment by fvictorio** [comment]
Oops, I thought this was just about compiling and didn't run `hh node` at the end. Sorry!

I was able to make it work by doing  `NODE_OPTIONS="--max-old-space-size=16000" hh node`, but that's a lot of memory, so there is indeed something wrong here. But, to avoid duplication, I'm going to close this issue in favor of #3471. I updated that issue with the relevant data from this one. Thanks for the great reproduction steps!
